pkg/cache/context.go

```diff
@@ -519,6 +524,38 @@ func (ctx *Context) IsPodFitNode(name string, node string) error {
 	return fmt.Errorf("predicates were not running because pod or node was not found in cache")
 }
 
+// call volume binder to bind pod volumes if necessary,
+// internally, volume binder maintains a cache (podBindingCache) for pod volumes,
+// and before calling this, they should have been updated by FindPodVolumes and AssumePodVolumes.
+func (ctx *Context) BindPodVolumes(pod *v1.Pod) error {
+	if assumedPod, exist := ctx.schedulerCache.GetPod(string(pod.UID)); exist {
+		// assumePodVolumes is called when we do assumePod, why we call it again here?
+		// ideally this should not happen, we can simply do assumePodVolumes, based on the return value,
+		// see if pod needs the binding operation.
+		// However, the assumePodVolumes was done in scheduler-core, because these assumed pods are cached
+		// during scheduling process as they have directly impact to other scheduling processes.
+		// The actual binding happens in the shim right before binding to a node.
+		allBound, err := ctx.volumeBinder.Binder.AssumePodVolumes(assumedPod, assumedPod.Spec.NodeName)
+		if err != nil {
+			log.Logger.Debug("AssumePodVolumes failed",
+				zap.String("podName", pod.Name),
+				zap.Error(err))
+			return err
+		}
+
+		if allBound {
```

Comment:
I don't think this is correct, we should only call this once in the `AssumePod` method.
The result  should be stored in the `assumedPods` mapping and leveraged here. 
```
if allBound, ok := assumedPods[pod id]; ok {
    if !allBound {
        return bind result volumes
    }
    return nil
}
return an error pod not assumed
```

Reply:
Unfortunately, this is the only option I come up with so far.
The reason I have commented in the code, please let me know if it makes sense

```
assumePodVolumes is called when we do assumePod, why we call it again here?
ideally this should not happen, we can simply do assumePodVolumes, based on the return value, see if pod needs the binding operation.
However, the assumePodVolumes was done in scheduler-core, because these assumed pods are cached during scheduling process as they have directly impact to other scheduling processes. The actual binding happens in the shim right before binding to a node.
```

Reply:
we're calling `AssumePodVolumes` twice on the same object:
1. first call in `AssumePod` in the context we deep copy the pod and then call `AssumePodVolumes`. After this we store the pod in the cache by calling `cache.AssumePod`. The cached version is thus the modified one.
1. the second call is in `BindPodVolumes` in the context, we pull the pod from the cache and call `AssumePodVolumes` again on the same object.

The two pods should thus be the same objects unless we keep overwriting the cached version which I would find strange and consider a bug. That is why I think we need to only call it once. The K8 scheduler also only calls it once.
It does require us to not ignore the result of the first `AssumePodVolumes` and pass it in to the `context.AssumePod` to set the boolean in cache assumedPods map.

Reply:
Make sense, fixed in latest PR.